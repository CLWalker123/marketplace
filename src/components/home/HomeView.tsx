import React, { useEffect } from 'react';
import { View, ScrollView, StyleSheet, ActivityIndicator } from 'react-native';
import { connect } from 'react-redux';
import Colors from '../../theme/colors';
import { ReduxState } from '../../redux/rootReducer';
import { Item } from '../item/types';
import { fetchAllItems as fetchAllItemsAction } from '../item/actions';
import { itemErrorSelector, itemPendingSelector, itemsSelector } from '../item/selectors';
import HomeEmptyView from './HomeEmptyView';
import HomeItemList from './HomeItemList';

interface Props {
  items: Record<string, Item>;
  itemError: string | null;
  itemPending: boolean;
  fetchAllItems: any;
}

export const HomeView = ({
  items,
  itemError,
  itemPending,
  fetchAllItems,
}: Props) => {

  useEffect(() => {
    fetchAllItems();
  }, [fetchAllItems])

  if ((!itemPending && itemError) || !Object.keys(items).length) {
    return <HomeEmptyView />
  }

  return (
    <View style={styles.container}>
      {itemPending ?
        <ActivityIndicator />
      :
        <HomeItemList items={items} />
      }
    </View>
  )
}

const mapStateToProps = (state: ReduxState) => {
  return {
    itemPending: itemPendingSelector(state),
    itemError: itemErrorSelector(state),
    items: itemsSelector(state),
  }
}

const mapDispatchToProps = {
  fetchAllItems: fetchAllItemsAction,
}

export default connect(mapStateToProps, mapDispatchToProps)(HomeView);

const styles = StyleSheet.create({
  container: {
    flex: 1,
  }
})