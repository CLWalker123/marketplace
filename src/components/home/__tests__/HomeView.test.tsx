import React from 'react';
import { shallow } from 'enzyme';
import { HomeView } from '../HomeView';

const mockState = {
  item: {
    pending: false,
    error: null,
    items: {},
  }
};

const mockItems = {
  1: {
    id: "1",
    name: "TV",
    price: 400000,
    time_remaining: 18427,
    views: 54,
  },
}

describe('HomeView', () => {
  it('renders the HomeEmptyView when no items are found', () => {
    const wrapper = shallow(
      <HomeView
        itemPending={mockState.item.pending}
        itemError={mockState.item.error}
        items={mockState.item.items}
        fetchAllItems={() => {}}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('renders the HomeView correctly when there is at least one item', () => {
    const wrapper = shallow(
      <HomeView
        itemPending={mockState.item.pending}
        itemError={mockState.item.error}
        items={mockItems}
        fetchAllItems={() => {}}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});