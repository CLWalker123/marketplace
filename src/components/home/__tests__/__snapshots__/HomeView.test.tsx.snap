// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeView renders the HomeEmptyView when no items are found 1`] = `<HomeEmptyView />`;

exports[`HomeView renders the HomeView correctly when there is at least one item 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <HomeItemList
    items={
      Object {
        "1": Object {
          "id": "1",
          "name": "TV",
          "price": 400000,
          "time_remaining": 18427,
          "views": 54,
        },
      }
    }
  />
</View>
`;
